### PR TITLE
Add show show_in_rest param to register_sidebar

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/53915

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:174]:
```suggestion
		return ! empty( $sidebar['show_in_rest'] );
```

This can be simplified to.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:145]: Why do we remove this check?

**Comment by TimothyBJacobs** [comment]
Won't we need a change in the widgets controller to check if a widget is part of a public sidebar?

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:145]: It was remove the `get_item_permissions_check` method, where always should have beenm
https://github.com/WordPress/wordpress-develop/pull/1578/files#diff-d5cf0688236a671f97ce276bb4900a3887b9ee1d2b3f8d4100fe72ad3797d13eR149-R151

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:174]: Don't we need to check it the key `show_in_rest` exists first?

**Comment by spacedmonkey** [comment]
@TimothyBJacobs Updated the PR with your feedback. 

One question. Should we / can we change raw instance values to only display in edit context. 

https://github.com/WordPress/wordpress-develop/blob/eb6b517aec3e04b1868a4ea4e4eb01a6a08a3751/src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php#L783-L787

I am wonder about sensitive data leaking?

**Comment by TimothyBJacobs** [comment]
Thanks. Looks like the test failure may be legit.

We would need to change the whole `instance` value to be only be exposed in an `edit` context, since the encoded form is just encoded, not hidden.

I think sensitive data leakage is definitely something we should be concerned about. While a theme author may opt-in to having their sidebar available, they might not know that a widget's setting contains an API key for instance.

However, I think there is definitely value in having the instance values available, not just the rendered widget, to allow for rendering the widget yourself.

I think for now we could omit the instance entirely, but it may make sense in the future to allow for `show_instance_in_rest` to be overloaded to a configuration array similarly to `register_meta` to allow for a widget author to selectively expose certain fields.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (TimothyBJacobs) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php]: No, `empty` won't error out on an undefined field.

**Comment by spacedmonkey** [comment]
Seems like the issue here was running `retrieve_widgets` twice, once in permission check and once in *_items function. I added a workaround for this, with some state is seems to work. I would love some feedback on managing state like this better. Thoughts @TimothyBJacobs

**Comment by TimothyBJacobs** [comment]
Pinging @adamziel as he's been looking into `retrieve_widgets` issues.

**Comment by adamziel** [comment]
> I added a workaround for this, with some state is seems to work. I would love some feedback on managing state like this better.

@spacedmonkey [I think the mental model of `retrieve_widgets` is broken](https://core.trac.wordpress.org/ticket/53816) right now. The name is confusing, that function doesn't just retrieve stuff, it actually updates the database. It's like "switching themes breaks the stored data, so we need to fix it every time we use it". I would like to make it more "we never break the data in the first place" but that's a larger project. For now I just [proposed a more accurate name](https://github.com/WordPress/wordpress-develop/pull/1524) and it has `commit` label so you might want to update this PR once that change is merged.

Since your issue seems to come from running that function twice and [we've seen similar problems](https://github.com/WordPress/gutenberg/issues/33335#issuecomment-879903958), I wonder if [this fix](https://github.com/WordPress/wordpress-develop/pull/1525) might help you. 

Other than that, I don't think that we have a good way around running that function multiple times like you do now. Well, maybe aside of maybe a hook that would run when a route is matched – that way we could have only a single call that would cover both the permissions check and the actual handler.

**Comment by spacedmonkey** [comment]
@adamziel Do you think that https://github.com/WordPress/wordpress-develop/pull/1525 blocks this PR? Would it effect anything in your PR? Did you review this PR?

**Comment by spacedmonkey** [comment]
> Also, one more question – should this PR be moved to the Gutenberg repo so that 5.7+plugin may benefit from these changes? That's what we did with WordPress/gutenberg#34230, it may or may not make sense here as well

Yeah, it can be backported.

**Comment by adamziel** [comment]
If it takes a few days then it would be convenient – potentially one check less in your PR. If it takes more, it's probably not worth it – we could re-evaluate that check at some later time.

**Comment by adamziel** [comment]
> Yeah, it can be backported.

Why not Gutenberg first? It would be picked up during the next major WP version as a part of the regular workflow. See the discussion in https://github.com/WordPress/gutenberg/issues/33810

**Comment by adamziel** [comment]
> Yeah, it can be backported.

Why not Gutenberg-first? These "double" changes have been challenging lately, see the discussion in  https://github.com/WordPress/gutenberg/issues/33810 and https://github.com/WordPress/gutenberg/pull/34536.

The sidebars endpoint is getting removed from the next Gutenberg version, so no problem there. It's all about having a consistent widgets endpoint now.

**Comment by spacedmonkey** [comment]
> Why not Gutenberg-first? These "double" changes have been challenging lately

This change doesn't make sense to backport. It has little use in the gutenberg plugin and it an opt in feature for devs. It is most useful for headless applications of WordPress.

**Comment by adamziel** [comment]
> This change doesn't make sense to backport. It has little use in the gutenberg plugin and it an opt in feature for devs. It is most useful for headless applications of WordPress.

I see, thank you for explaining! I am only concerned about merging changes later on. If we keep updating the same endpoints both in core and in the Gutenberg repo, we may end up with conflicting changes. On the other hand, I think I discussed that with @gziolo in https://github.com/WordPress/gutenberg/issues/33810 and the conclusion was that in practice this rarely becomes a problem.

**Comment by spacedmonkey** [comment]
> this rarely becomes a problem so maybe we're good here?

If there was any value in adding this to the gutenberg plugin, I would have done it both places. I will do that future PRs.

**Comment by spacedmonkey** [comment]
Unit tests are passing!

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php
Comment (adamziel) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php:488]: I'd add `@see retrieve_widgets` to explain what "lost" means

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (adamziel) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:287]: I'd add `@see retrieve_widgets` to explain what "lost" means

File: tests/phpunit/tests/rest-api/rest-widgets-controller.php
Comment (adamziel) [inline tests/phpunit/tests/rest-api/rest-widgets-controller.php:243]: I'm confused because the test name says "no permission public", but then it sets up a sidebar with two widgets and the endpoint returns both. I'd add a few test cases where the API returns a limited list of widgets or a 403 error.

File: tests/phpunit/tests/rest-api/rest-sidebars-controller.php
Comment (adamziel) [inline tests/phpunit/tests/rest-api/rest-sidebars-controller.php:170]: See https://github.com/WordPress/wordpress-develop/pull/1578/files#r719494361

File: src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-sidebars-controller.php:287]: Fixed.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php
Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-widgets-controller.php:488]: Fixed

File: tests/phpunit/tests/rest-api/rest-sidebars-controller.php
Comment (spacedmonkey) [reply tests/phpunit/tests/rest-api/rest-sidebars-controller.php:201]: Test added!

File: tests/phpunit/tests/rest-api/rest-widgets-controller.php
Comment (spacedmonkey) [reply tests/phpunit/tests/rest-api/rest-widgets-controller.php:285]: Test added!

**Review by adamziel — APPROVED** [review]
LGTM

**Comment by TimothyBJacobs** [comment]
Merged in https://core.trac.wordpress.org/changeset/52016.

**Pull request closed without merging**
